### DISCUSSION TITLE
Support a way to have `nameof` refer to name of containing method.

### DISCUSSION THREAD

**Discussion by The-Futurist** [dotnet/csharplang#5289] [Language Ideas] [answered]
Sometimes it would be helpful to be able to write:

```cs
Support.RunSQL($"Create{nameof(ProcedureWrapper_Test_4)}");
```
as this:

```cs
Support.RunSQL($"Create{nameof()}");
```

or as this

```cs
Support.RunSQL($"Create{methodname}");
```
or something, this comes up a lot when writing repetitive code like unit tests where many tests use the name of the containing method in some way.

It's easy to copy/paste these tests as the test suite develops and easy to end up with some `TestMethod_12` using `nameof(TestMethod_11)` for example.

Writing name() might be a reasonable idea here because it could not be a breaking change...

Thoughts?

**Comment by jnm2** [comment, marked as answer]
As of C# 5, this can be implemented using CallerMemberNameAttribute:

```cs
void ProcedureWrapper_Test_4()
{
    Support.RunSQL($"Create{CurrentMemberName()}");
}

static string CurrentMemberName([CallerMemberName] string callerMemberName = null) => callerMemberName;
```

To avoid `using static` and avoid declaring a method in every class where this is used, a helper class could enable the syntax `CurrentMember.Name()`:

```cs
internal static class CurrentMember
{
    public static string Name([CallerMemberName] string callerMemberName = null) => callerMemberName;
}
```

**Reply by The-Futurist** [reply]
This is good, I seem to recall hearing of that attribute but hardly ever looked at it, this is a neat solution!
